### PR TITLE
Lloyd relaxation (+ android compatibility)

### PR DESCRIPTION
**Pull request by kjav**
Hi, I've added lloyd relaxation to the module (https://en.wikipedia.org/wiki/Lloyd%27s_algorithm). I have also replaced the "long" pointer casts in the module as they did not allow the module to compile on 32 bit architectures such as armv7.

### CODE REVIEW

File: voronoi.cpp [hunk 1]
```diff
@@ -139,6 +139,44 @@ void userfree(void* ctx, void* ptr) {
 	return memfree(ptr);
 }
 
+void Voronoi::relax_points(int iterations = 1) {
+	std::vector<jcv_point> new_points;
+	new_points.reserve(_points.size());
+	for (int j = 0; j < iterations; j++) {
+		Ref<VoronoiDiagram> diagram { memnew(VoronoiDiagram) };
```
Comment (rakai93) [inline voronoi.cpp]: As far as I see, the creation of a godot class is not needed, is it? Below, you only make use of the jcv* functions and only operate on diagram->_diagram, not using any methods of the godot classes. Thus, I think using a jcv_diagram directly here (as stack variable as it is anyways only used in this scope) is the easiest solution with the least overhead.

**Review by rakai93 — CHANGES_REQUESTED** [review]
Very cool contribution! But I think currently a lot of unnecessary overhead is done, see comment.

File: voronoi.cpp (hunk 1 above)
Comment (kjav) [reply voronoi.cpp]: Thanks so much for the feedback - I'm not a c/c++ developer so I don't have a good intuition of these things. I've attempted to resolve your issue, please see my latest commit :)

File: voronoi.cpp [hunk 2]
```diff
@@ -143,19 +143,18 @@ void Voronoi::relax_points(int iterations = 1) {
 	std::vector<jcv_point> new_points;
 	new_points.reserve(_points.size());
 	for (int j = 0; j < iterations; j++) {
-		Ref<VoronoiDiagram> diagram { memnew(VoronoiDiagram) };
-		jcv_diagram_generate_useralloc(
+		jcv_diagram diagram;
+		memset(&diagram, 0, sizeof(jcv_diagram));
+		jcv_diagram_generate(
```
Comment (rakai93) [inline voronoi.cpp:148]: I think using jcv_diagram_generate_useralloc was the correct move here, as it internally still allocates some memory.

File: voronoi.cpp [hunk 3]
```diff
@@ -143,19 +143,18 @@ void Voronoi::relax_points(int iterations = 1) {
...
 			_points.size(),
 			_points.data(),
-			_boundaries.get(),
-			NULL,
-			&useralloc,
-			&userfree,
-			diagram->_diagram
+			0,
```
Comment (rakai93) [inline voronoi.cpp:151]: Using NULL here was a little more correct (Sorry for being picky, but if there is a constant meant for that we should use it :))

File: voronoi.cpp [hunk 4]
```diff
@@ -36,7 +36,7 @@ Variant VoronoiSite::edges() const {
 	Vector<Variant> result;
 	const jcv_graphedge* graphedge = _site->edges;
 	while (graphedge) {
-		result.push_back(_diagram->_edges_by_address[Variant((long)graphedge->edge)]);
+		result.push_back(_diagram->_edges_by_address[Variant((uint64_t)graphedge->edge)]);
```
Comment (rakai93) [inline voronoi.cpp:39]: I actually pushed some more commits to master from myself where I fix some compiler warnings on MSVC, including the plain conversion from pointers to long. In 6b9b1b7ab9b01e377098c1cfe62df555d3f30038, I replaced this C style cast with something that is supposed to be platform-neutral. Could you maybe pull the master branch into your master branch and check if my commit also works fine on Android? :)

**Review by rakai93 — CHANGES_REQUESTED** [review]

**Comment by rakai93** [comment]
@kjav I created an own branch with your commits and resolved the merge conflicts with master: #4 I wanted to use it myself, so I created it right away. Can you check if this works for you? :) If so, we can merge this one instead.

**Comment by rakai93** [comment]
I merged PR #3 as it also fixes one more huge bug in relax_points but also re-adds compatibility with the master branch, e.g. C++17 stuff. I hope you can use the master HEAD now for your project!

**Comment by kjav** [comment]
Thanks so much! I am on holiday this week but will follow up when I test
it. Thanks again for the c++ advice.

**Pull request merged**